src/foam/parse/parse.js

```diff
@@ -602,6 +655,59 @@ foam.CLASS({
   ]
 });
 
+foam.CLASS({
+  package: 'foam.parse',
+  name: 'UntilEscaped',
+  extends: 'foam.parse.ParserDecorator',
+
+  documentation: `Matches any characters until the terminating pattern,
+    unless terminating pattern has a leading escape pattern.
+    Consumes and discards the terminating pattern when found.  Fails if termination was never found.`,
+
+  properties: [
+    {
+      name: 'esc',
+      class: 'foam.parse.ParserProperty',
+      final: true
+    },
+    {
+      name: 'term',
+      class: 'foam.parse.ParserProperty',
+      final: true
+    }
+  ],
+
+  methods: [
+    function parse(ps, obj) {
+      var ret = [];
+      var esc = this.esc;
+      var term = this.term;
+
+      while ( ps.valid ) {
```

Comment:
I think you could compose this from existing primitives like so:

`repeat(not(<terminator>, alt(seq1(2, <escape>, <terminator>), anyChar))`

Although a shortcut for this could be useful, but instead of implementing parse() again you could just extend ProxyParser and put an expression/factory on delegate based upon the "terminator" and "escape" properties.

Reply:
Althought in Java we have foam.lib.json.StringParser which does a similar job and this has a written by hand parser, but that was somewhat done for performance reasons.

Reply:
Oh awesome, I was hoping there was a way to do that with the existing parsers - thanks!

Reply:
I actually need to capture my escape symbol as well (I have an action that returns the character escaped for valid JSON escapes), so I've tried

`repeat(not('"', alt(sym('json-escape'), anyChar())))`

but this is causing my parser to stop early. Any ideas?

Reply:
I just realized I don't need to add a parser at all - if I can construct untilEscaped with primitives then I can just add a method to foam.parse.Parsers that uses the primitives the same way as in a grammar, right?